### DISCUSSION TITLE
Local sync of subscriptions and playlists

### DISCUSSION THREAD

**Discussion by PhrozenByte** [libre-tube/LibreTube#5195] [Ideas]
### Describe your suggested feature

When using an authenticated Piped instance, subscriptions and playlists are just a mirror of the subscriptions and playlists stored on the instance's server. This means that if someone chooses to switch to another instance, subscriptions and playlists also switch to that of the new instance.

Unfortunately Piped instances aren't very reliable - even the most reliable instances (I personally use kavin.rocks and adminforge.de) have issues from time to time. That's no big deal in general because it's easy to switch to another instance, however, since subscriptions and playlists are stored on the instance's server, they get out of sync all the time. This is annoying. Thus I'm suggesting that LibreTube keeps a local copy of subscriptions and playlists and compares that to the remote list of subscriptions and playlists when authenticating with a new instance. If the lists differ (and only then), a simple dialogue is shown to ask the user what to do.

This dialogue might look like the following:

> Your locally stored channel subscriptions and those of the authenticated Piped instance differ. Do you want to upload your local subscriptions to the remote Piped instance, or do you want to use the remote Piped instance's subscriptions instead?
>
> [ Upload local subscriptions ]
> [ Use remote subscriptions ]

If playlists differ (including both whole playlists, and single videos within), another dialogue with the same question for playlists pops up.

I was thinking about whether LibreTube should also add some kind of a diff dialogue to select or de-select channels individually, but this is likely overkill. A simple "use local" or "use remote" button is fine.

### Other details

I'm aware that one can already export and import subscriptions and playlists. However, keep in mind that this requires the instance to be up and running. If the instance is down - which is likely the case since that's the reason why I want to switch the instance in the first place - one can't export subscriptions and playlists. Adding a local sync of subscriptions and playlists also enables exporting them when the instance is down (even though a warning should be shown then).

**Related suggestion / optional addition to this suggestion**

This might be combined with LibreTube's existing local subscriptions and playlists feature (i.e. when the user isn't using an authenticated Piped instance). By "combining" these two features I mean that LibreTube isn't actually creating a distinct copy of remote subscriptions and playlists for syncing purposes, but "just" keeps local and remote subscriptions and playlists in sync. Remote subscriptions and playlists always take precedence unless the user actively switches instances. This would also solve #4877. Another major advantage of combining this with the local subscriptions and playlists feature is that one doesn't actually have to switch to another authenticated Piped instance: One can simply just log-out and use the local-only implementation until the Piped instance is back online.

**Related issues**

- #4877 as explained earlier
- #3144 wants to implement an auto backup feature; this is a whole different story, because it wouldn't affect switching instances and also includes settings and local bookmarks
- #1633 (and probably some more) reported this currently expected behaviour as bug

**Sharing some :heart:**

Keep up your amazing work! I switched to LibreTube some years ago out of despair, but LibreTube gets better and better with every single release. It's already way better than the original YouTube app, thank you so much for your work! :heart: 

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I will fill out all of the requested information in this form.

**Comment by voidplayer** [comment]
This is what i miss the most in the app rn

I was looking into setup my own auth instance because my own server is very small and always online so i can simply jump instances, but this is even better

A local copy of your subs if the auth instance is down. I wont be able to add new ones, but at least it will work
